google-cloud-spanner/src/test/java/com/google/cloud/spanner/spi/v1/GfeLatencyTest.java

```diff
@@ -0,0 +1,346 @@
+                                        return serverCallHandler.startCall(
+                                                new ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT>(serverCall) {
+                                                    @Override
+                                                    public void sendHeaders(Metadata headers){
+                                                        headers.put(
+                                                                Metadata.Key.of("server-timing", Metadata.ASCII_STRING_MARSHALLER),
+                                                                String.format("gfet4t7; dur=%d", fakeServerTiming.get()));
+                                                        super.sendHeaders(headers);
+                                                    }
+                                                }, headers);
+                                    }
+                                })
+                        .build()
+                        .start();
+        optionsMap.put(SpannerRpc.Option.CHANNEL_HINT, 1L);
+        spanner = createSpannerOptions(address,server).getService();
+        databaseClient =
+                spanner.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+
+
+        mockSpannerNoHeader = new MockSpannerServiceImpl();
+        mockSpannerNoHeader.setAbortProbability(0.0D);
+        mockSpannerNoHeader.putStatementResult(MockSpannerServiceImpl.StatementResult.query(SELECT1AND2, SELECT1_RESULTSET));
+        mockSpannerNoHeader.putStatementResult(MockSpannerServiceImpl.StatementResult.update(UPDATE_FOO_STATEMENT, 1L));
+        addressNoHeader = new InetSocketAddress("localhost", 1025);
+        serverNoHeader =
+                NettyServerBuilder.forAddress(addressNoHeader)
+                        .addService(mockSpannerNoHeader)
+                        .build()
+                        .start();
+        spannerNoHeader = createSpannerOptions(addressNoHeader, serverNoHeader).getService();
+        databaseClientNoHeader =
+                spannerNoHeader.getDatabaseClient(DatabaseId.of(projectId, instanceId, databaseId));
+    }
+
+    @AfterClass
+    public static void stopServer() throws InterruptedException {
+        if (spanner != null) {
+            spanner.close();
+            server.shutdown();
+            server.awaitTermination();
+        }
+
+        if(spannerNoHeader != null){
+            spannerNoHeader.close();
+            serverNoHeader.shutdown();
+            serverNoHeader.awaitTermination();
+        }
+    }
+
+    @After
+    public void reset() {
+        mockSpanner.reset();
+        mockSpannerNoHeader.reset();
+    }
+
+    @Test
+    public void testGfeLatencyExecuteStreamingSql() throws InterruptedException {
+        try(ResultSet rs = databaseClient.singleUse().executeQuery(SELECT1AND2)){
+            rs.next();
+        }
+        Thread.sleep(WAIT_FOR_METRICS_TIME_MS);
```

Comment:
We should not sleep, this creates a barrier on how fast our tests can be. Is there any other way we could do this?

Reply:
I've changed the code to take a maximum number of retries, sleeping for 1_000 milliseconds between each retry. 
However, I still had to sleep the thread while checking for the header missing count (when expected value is 1) so as to allow the previous 0 to be rewritten. Please let me know if this is alright.